lamindb/dev/db/_add.py

```diff
@@ -176,32 +176,35 @@ def local_instance_storage_matches_local_parent(dobject: DObject):
     return str(storage.root) in parents
 
 
-def write_objectkey(record) -> None:
+def get_storage_root_str():
+    root = setup_settings.instance.storage.root
+    root_str = str(root)
+    if isinstance(root, UPath):
+        root_str = root_str.rstrip("/")
+    return root_str
+
+
+def write_objectkey(record: sqm.SQLModel) -> None:
     """Write to _objectkey.
 
     An objectkey excludes the storage root and the file suffix.
     """
-    storage = setup_settings.instance.storage
 
-    def set_objectkey(record: sqm.SQLModel, filepath: Union[Path, UPath]):
-        if not isinstance(filepath, UPath):  # is local filepath
-            filepath_str = filepath.resolve().as_posix()
-        else:  # is remote path
-            filepath_str = filepath.as_posix()
-        root_str = storage.root.as_posix()
-        if root_str[-1] != "/":
-            root_str += "/"
-
-        if not isinstance(record, (DObject, DFolder)):
-            raise NotImplementedError
-        # for DObject, _objectkey is relative path without suffix
+    def set_objectkey(record: Union[DObject, DFolder], filepath: Union[Path, UPath]):
+        root_str = get_storage_root_str()
+
+        try:  # is local filepath
+            relpath = filepath.relative_to(root_str)
+        except ValueError:  # is cloud filepath
+            # remove the server prefix
+            relpath = Path(filepath.as_posix().replace(root_str, ""))
+
+        # for DObject, _objectkey is relative path to the storage root without suffix
         _objectkey = (
-            filepath_str.replace(root_str, "").replace(record.suffix, "")
-            if isinstance(record, DObject)
-            else filepath_str.replace(root_str, "")
+            relpath.parent / record.name if isinstance(record, DObject) else relpath
         )
 
-        set_attribute(record, "_objectkey", _objectkey)
+        set_attribute(record, "_objectkey", str(_objectkey))
```

Comment:
I think it is better to use `_objectkey.as_posix()` instead of `str(_objectkey)` when writing to the db, because on Windows and Linux\Mac `str(_objectkey)` gives different results (\\ vs /).

Reply:
Done!